### PR TITLE
Add end to end tests for Generated Go Code

### PR DESCRIPTION
**Pull request by DomBlack**
This commit adds to the existing daemon run tests to
also exercise go clients against the "echo" test app.
This gives us a full end to end test of all generated Go
code from the clients into the server and back.

These tests attempt to exercise all the different ways
Encore supports API's being written, with all the supported
parameter types - attempting to catch any edge cases.

It's my intention to add in a separate PR the same client
tests, but in Typescript for the Typescript generated clients.

These tests already picked up a couple of bugs, which I've fixed in this PR

### CODE REVIEW

File: cli/daemon/run/run_test.go [hunk 1]
```diff
@@ -88,211 +97,235 @@ func TestStartProc(t *testing.T) {
...
+		{
+			w := httptest.NewRecorder()
+			req := httptest.NewRequest("GET", "/echo.AppMeta", nil)
+			run.ServeHTTP(w, req)
+			c.Assert(w.Code, qt.Equals, 200)
+
+			// we need to extract the API Base URL as it will change due to the `RuntimePort: 0` above
+			bytes := w.Body.Bytes()
+			got := make(map[string]string)
+			_ = json.Unmarshal(bytes, &got)
```
Comment (eandre) [inline cli/daemon/run/run_test.go:307]: Is it guaranteed all values are strings or should this be `map[string]any`?

Comment (DomBlack) [reply cli/daemon/run/run_test.go:307]: 🤷 This was the original test code, I've not modified it apart from tabbing it in to run as a sub test.

Comment (DomBlack) [reply cli/daemon/run/run_test.go:307]: Yes it is, this is the response type in the test app

```
type AppMetadata struct {
	AppID      string
	APIBaseURL string
	EnvName    string
	EnvType    string
}
```

File: cli/daemon/run/run_test.go [hunk 2]
```diff
@@ -352,3 +385,27 @@ func (l testRunLogger) runStderr(r *Run, line []byte) {
 	line = bytes.TrimSuffix(line, []byte{'\n'})
 	l.t.Log(string(line))
 }
+
+func proxyTcp(ctx context.Context, ln net.Listener, client *yamux.Session) {
+	for ctx.Err() == nil {
+		conn, err := ln.Accept()
```
[the remark this answers is not in the document]
Comment (DomBlack) [reply cli/daemon/run/run_test.go:391]: The listener is closed at the top of `TestEndToEndWithApp` with a defer statement, which will cause the call to `Accept()` to error. The if statement checking for `use of closed network connection` prevents that error being reported as test failure

File: cli/internal/codegen/testdata/expected_golang.go [hunk 3]
```diff
@@ -402,6 +402,9 @@ func (c *svcClient) Webhook(ctx context.Context, a string, b string, request *ht
 		return nil, fmt.Errorf("unable to parse api url: %w", err)
 	}
 	request = request.WithContext(ctx)
+	if request.Method == "" {
+		request.Method = "POST"
+	}
```
[the remark this answers is not in the document]
Comment (DomBlack) [reply cli/internal/codegen/testdata/expected_golang.go:409]: We override the URL object passed to a raw endpoint already as that's set at the API level. So we don't expect request objects passed in to have the URL set (and if they are, we ignore them).

The same applies for the method, the API helper function I'm using in the client knows how to call the raw endpoint, so I expect it to be able to call said endpoint with what ever request object provide.

[the remark this answers is not in the document]
Comment (DomBlack) [reply cli/internal/codegen/testdata/expected_golang.go:409]: i.e. we should be able to make a request like this:

https://github.com/encoredev/encore/pull/229/files#diff-cd3cd9de3bee67c24f7159ff7ece28f7ce5b0b088c2dce350d337dd9d49e4d4bR178-R185

[the remark this answers is not in the document]
Comment (eandre) [reply cli/internal/codegen/testdata/expected_golang.go:409]: Well that’s the case because we’re providing the convenience of accepting path parameters in the signature.

But method seems different: in valid Go code it’s not really possible to create an `*http.Request` without a valid method, is it? There are lots of field we expect you to have set on the Request you pass us, why does method deserve special treatment?

I guess my question is: what was the scenario that motivated this? And was that code valid or invalid?

[the remark this answers is not in the document]
Comment (eandre) [reply cli/internal/codegen/testdata/expected_golang.go:409]: Should we accept just a raw object constructed like that, though? The standard constructor for outgoing requests does a whole lot more: https://cs.opensource.google/go/go/+/refs/tags/go1.18.2:src/net/http/request.go;l=856

[the remark this answers is not in the document]
Comment (DomBlack) [reply cli/internal/codegen/testdata/expected_golang.go:409]: > I guess my question is: what was the scenario that motivated this? And was that code valid or invalid?

In this example:

```go
		req := &http.Request{
			Header: http.Header{
				"X-Test-Header": []string{"test"},
			},
			Body: io.NopCloser(strings.NewReader("this is a test body")),
		}

		rsp, err := api.Test.RawEndpoint(ctx, "hello", req)
```

A _valid_ request object is constructed, with a header and body provided - with is what the Raw endpoint needs in this case. Our generated client will add anything else needed to call that endpoint, including;
- Auth token to the header
- The URL to the handler
- The method required by the handler

Asking the user to include the method or URL requests in the question why are they even using the generated client in that case?

It's also worth noting the documentation for `http.Header` states method can be empty - and will be assumed to be `GET` in that case. In our case; we're actually assuming it to be the method required by that API.

[the remark this answers is not in the document]
Comment (DomBlack) [reply cli/internal/codegen/testdata/expected_golang.go:409]: > Should we accept just a raw object constructed like that, though? The standard constructor for outgoing requests does a whole lot more

I'm conflicted here; I'd argue yes - because we can populate the rest.

We could create a `api.Test.NewRawEndpointRequest(body io.Reader) *http.Request` method, which we then expect you to use with the `api.Test.RawEndpoint(ctx context.Context, id string, req *http.Request)` method.

But then;
1. How could we easily support path params without the using having to do the encoding themselves?
2. Should things like the auth token be added to that returned request object
3. If we return that, we still want the request to be used through the API client, so other options in the baseClient can be applied - having a raw req object though would be tempting to simply execute it using any round tripper avaliable.

[the remark this answers is not in the document]
Comment (eandre) [reply cli/internal/codegen/testdata/expected_golang.go:409]: Okay, since `(*http.Request).Method`'s documentation specifies it can be empty that's a compelling reason to support a default value. But I don't think we should be making "smart decisions" about sensible defaults for raw requests, because the whole point of raw requests is to enable full control over the request. So we should be clear about which fields we override or support empty values for. I don't think we should document that you can pass in a raw `&http.Request{Header: foo, Body: bar}` literal, because that's in general not a supported way of using `(http.RoundTripper).Do` as far as I know.

File: cli/internal/codegen/testdata/expected_typescript.ts [hunk 4]
```diff
@@ -326,8 +326,8 @@ export namespace svc {
             return await resp.json() as Tuple<boolean, Foo>
         }
 
-        public async Webhook(a: string, b: string, body?: BodyInit, options?: CallParameters): Promise<Response> {
-            return this.baseClient.callAPI("POST", `/webhook/${a}/${b}`, false, body, options)
+        public async Webhook(method: "*", a: string, b: string, body?: BodyInit, options?: CallParameters): Promise<Response> {
+            return this.baseClient.callAPI(method, `/webhook/${a}/${b}`, false, body, options)
```
Comment (eandre) [inline cli/internal/codegen/testdata/expected_typescript.ts:330]: This should be `method: string` in case we allow any method I guess

**Review by eandre — APPROVED** [review]
